core/src/main/java/io/grpc/internal/CompositeReadableBuffer.java

```diff
@@ -202,32 +234,11 @@ private void advanceBufferIfNecessary() {
   }
 
   /**
-   * A simple read operation to perform on a single {@link ReadableBuffer}. All state management for
-   * the buffers is done by {@link CompositeReadableBuffer#execute(ReadOperation, int)}.
+   * A simple read operation to perform on a single {@link ReadableBuffer}.
+   * All state management for the buffers is done by
+   * {@link CompositeReadableBuffer#execute(ReadOperation, int, Object, int)}.
    */
-  private abstract static class ReadOperation {
-    /**
-     * Only used by {@link CompositeReadableBuffer#readUnsignedByte()}.
-     */
-    int value;
-
-    /**
-     * Only used by {@link CompositeReadableBuffer#readBytes(OutputStream, int)}.
-     */
-    IOException ex;
-
-    final void read(ReadableBuffer buffer, int length) {
-      try {
-        value = readInternal(buffer, length);
-      } catch (IOException e) {
-        ex = e;
-      }
-    }
-
-    final boolean isError() {
-      return ex != null;
-    }
-
-    abstract int readInternal(ReadableBuffer buffer, int length) throws IOException;
+  private interface ReadOperation<T> {
+    int read(ReadableBuffer buffer, int length, T dest, int value) throws IOException;
```

Comment:
Somewhere the relationship between `value` and the returned `int` needs to be mentioned, since that might be surprising and doesn't "jump out" obviously in the code.